### PR TITLE
Naive implementation of loss functions defined in GE2E Loss (Moreno, et. al.)

### LINKED ISSUE #5: implement loss function

### PR DESCRIPTION
**Pull request by pieceOpiland**
Hey all, there were two different approaches to the loss function defined in the [GE2E paper](https://arxiv.org/pdf/1710.10467.pdf): 

* Softmax Loss - Better for Text Independent Speaker Verification
* Contrast Loss - Better for Text Dependent Speaker Verification

I wound up taking a shot at both approaches to loss outlined in the GE2E paper, and figured we might be able to discuss which approach folks thought was best?

Contrast Loss appears to be a bit more agressive in the way it "pushes" matching embeddings away from incorrect centroids, whereas Softmax Loss seems to consider all centroids.

Closes #5

### CODE REVIEW

**Comment by perrychu** [comment]
I think we're going for text independent, so softmax loss probably makes sense as the since it performed better for text independent in the GE2E paper.

I'm a little puzzled why j=k is not excluded in the LogSumExp part of Softmax Loss (equation 6). I guess the idea is to push similarity with the other centroids to be less than the target speaker centroid, not necessarily push similarity with other centroids to be zero.

The CorentinJ repo went with a different implementation of loss altogether. I'm not sure how that compares with either softmax or contrastive.

**Comment by pieceOpiland** [comment]
If you go [back far enough](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/719315acda723b5da99f560fab26b4c0e8d66321/SV2TTS/model.py#L86-L90), you can see where a non-vectorized version of the softmax loss is being used. I'm not sure the source of the loss being used in that project now though.

**Comment by pieceOpiland** [comment]
The paper states, "We put a softmax on S_{ji,k} for k = 1, . . . , N that makes the output equal to 1 iff k = j, otherwise makes the output equal to 0.", so it sounds like there's no effect of including k = j in softmax loss.
